### PR TITLE
settings: Add allow_user_npmrc option

### LINKED ISSUE #19806: Add options for enable user npmrc
### Check for existing issues

- [X] Completed

### Describe the feature

It was disabled user npmrc and global npmrc in #3324  . Is there any options for enable user npmrc?

The user npmrc is useful in some case such as change npm registry to a localed mirror, or disable ssl verification behind a proxy network with custom CA.

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by yuekcc**
Closes #19806

Add `allow_user_npmrc` to `node` section. Set `true` to allow npm install packages with user's .npmrc. 

Settings looks like this:

```jsonc
{
  // ...
  "node": {
    "allow_user_npmrc": true
  },
  // ...
}
```

Release Notes:

- Added a `allow_user_npmrc` option to enable user's .npmrc instead of  a blank npmrc when npm installing packages.

### CODE REVIEW

File: crates/project/src/project_settings.rs [hunk 1]
```diff
@@ -67,6 +67,10 @@ pub struct NodeBinarySettings {
     /// If disabled, zed will download its own copy of node.
     #[serde(default)]
     pub ignore_system_version: Option<bool>,
+    /// Allow using user's npmrc
+    ///
+    /// default: false
```
Comment (maxdeviant) [inline crates/project/src/project_settings.rs:72]:
```suggestion
    /// Default: false
```

**Comment by ConradIrwin** [comment]
I think we don't want to pull in the users' npmrc for all settings, as it includes things that will break what we want to do (which is why we don't include it right now).

I think it would be reasonable to read the proxy settings (without adding a configuration option). 

I'm not sure about the registry setting. Can you please explain more about your NPM setup - would installing the packages we expect to be on the global registry work from your custom one?

**Comment by yuekcc** [comment]
Using the `npmrc` file mainly addresses the issue of utilizing an npm registry mirror.

I use Zed in my company's intranet environment, where access to the public npm registry is not possible. Therefore, we created an npm mirror, but with a custom certificate. With a blank configuration, npm enforces the default settings, making it impossible to ignore the certificate error.

My typical `npmrc` configuration looks like this:

```
strict-ssl = false
registry = "https://some-npm-registry-mirror.url"
```

This also applies to other special network environments, such as when accessing the npm registry from China, which can be very slow, and therefore usually requires configuring an npm registry mirror.

Since these configurations pertain to npm, I propose adding just one switch: `enable_user_npmrc`.

Additionally, if Zed could offer an offline plugin installation method, that would be even better—similar to how VSCode allows plugin installation via VSIX files.

**Comment by ConradIrwin** [comment]
We are working towards a world where if the required dependencies are available already we don't install them, but not quite there yet.

I like the idea of being able to provide a download that encompasses the extension and any required language server installations.

For now, let's (without config) read the values from the npmrc that make this work for you:

```
strict-ssl
registry
certificate ?
```

**Comment by yuekcc** [comment]
> For now, let's (without config) read the values from the npmrc that make this work for you:
>  ```
> strict-ssl
> registry
> certificate ?
> ```

That's good. 

I'll close this PR. Thank you.

**Pull request closed without merging**